File: javascript/49. Group Anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    if(strs.length === 1) return [strs]
    
    var sortedStrsMap = strs.reduce((map, str, idx) => {
        let sorted = str.split('').sort().join('')
        map[sorted] = map[sorted] || { val: [] }
        map[sorted].val.push(idx)
        return map
    }, {})
    
    return Object.keys(sortedStrsMap).map((key) => {
        return sortedStrsMap[key].val.map(idx => strs[idx])
    })
};

File: java/49. Group Anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> result = new ArrayList<>();
        if(strs.length == 1) {
            List<String> newCell = new ArrayList<>();
            newCell.add(strs[0]);
            result.add(newCell);
            return result;
        }

        Map<String, Integer> map = new HashMap<>();

        for(int i = 0; i < strs.length; i++) {
            char[] arr = strs[i].toCharArray();
            Arrays.sort(arr);
            String sorted = new String(arr);
            Object index = map.get(sorted);
            if(index != null) {
                result.get((int) index).add(strs[i]);
            } else {
                List<String> newCell = new ArrayList<>();
                newCell.add(strs[i]);
                result.add(newCell);
                map.put(sorted, result.size() - 1);
            }
        }
        return result;
    }
}